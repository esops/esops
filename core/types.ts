import {sanitizeComponent} from './lenses'
import {listTreeSync} from 'fs-plus'
import {Path} from 'path'
import {PathLike} from 'fs'

// TODO: Figure out how to import path as a module without this patch.
declare module 'path' {
  interface Path {
    normalize(p: string): string
    join(...paths: any[]): string
    resolve(...pathSegments: any[]): string
    isAbsolute(p: string): boolean
    relative(from: string, to: string): string
    dirname(p: string): string
    basename(p: string, ext?: string): string
    extname(p: string): string
    sep: string
    delimiter: string
    parse(p: string): ParsedPath
    format(pP: FormatInputPathObject): string
  }
}

export type ComponentVariables = {}

export type ComponentOptions = {
  mergeJson: []
  mergeFiles: []
  overwrite: []
  merge: boolean
  publish: boolean
  o: string
  out: string
}

export type Component =
  | string
  | [string, ComponentVariables?, ComponentOptions?]
  | SanitizedComponent

export type SanitizedComponent = [string, ComponentVariables, ComponentOptions]

export type Compose = Component | Component[]

export type SanitizedCompose = SanitizedComponent[]

type Prompt = boolean | Error

export type UserParams = {
  root: string
  destination?: string
  compose?: Compose
  logLevel?: string
  prompts?: Prompt[]
}

export type Effects = {
  filesystem: {
    path: Path
    appCache: {
      getRenderPrepFolder: () => Promise<string>
    }
    listTreeSync: (str: string) => string[]
    isDirectory: {
      sync: (isDirectory: string) => boolean
    }
    existsSync: (str: string) => boolean
    readFileSync(path: PathLike, options?: {encoding: string} | string): string
    writeFileSync: (s1: string, s2: string) => void
    forceCopy: (s1: string, s2: string) => void
  }
  resolve: (
    params: Params,
    sanitizeComponent: SanitizedComponent
  ) => SanitizedComponent
  ui: {
    getTabs: (n: number) => string
    info: (u: unknown) => void
  }
  shell: {}
  vm: {}
}

export type Params = {
  effects: Effects
  destination: string
  root: string
  parent: SanitizedComponent
  compose?: Compose
  logLevel: string
  prompts: Prompt[]
  results: Report[]
  treeDepth: number
}

export type EsopsConfigObject = {
  destination?: string
  compose?: Compose
}

export type EsopsConfig = EsopsConfigObject | Compose

export type Manifest = {
  from: string
  to: string
  relativePath: string
  shouldMergeJson: boolean
  shouldMergeFile: boolean
  shouldGitPublish: boolean
  mergeJsonArrays: string[]
}

export type Actions = Manifest[]
export type Report = Manifest[]

export type ResolverFunc = (
  params: any,
  sanitizedComponent: any
) => Promise<string>

export type ResolverExtension = {
  is: (componentString: string) => boolean
  resolve: ResolverFunc
  COMPONENT_TYPE: string
}
