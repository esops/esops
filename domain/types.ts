type Path = string

export type ComponentVariables = {}

export type ComponentOptions = {
  mergeJson: []
  mergeFiles: []
  overwrite: []
}

export type Component =
  | string
  | [string, ComponentVariables]
  | SanitizedComponent

export type SanitizedComponent = [string, ComponentVariables, ComponentOptions]

export type Compose = Component | Component[]

type Prompt = boolean | Error

export type UserParams = {
  root: string
  destination?: string
  compose?: Compose
  logLevel?: string
  prompts?: Prompt[]
}

export type Effects = {
  filesystem: {
    appCache: {
      getRenderPrepFolder: () => Promise<string>
    }
  }
  resolver: {}
  ui: {}
  shell: {}
  vm: {}
}

export type Params = {
  effects: {}
  destination: string
  root: string
  parent: SanitizedComponent
  compose?: Compose
  logLevel: string
  prompts: Prompt[]
  results: Report[]
  treeDepth: number
}

export type EsopsConfigObject = {
  destination?: Path
  compose?: Compose
}

export type EsopsConfig = EsopsConfigObject | Compose

export type Report = {
  from: string
  to: string
  relativePath: string
  shouldMergeJson: boolean
  shouldMergeFile: boolean
  shouldGitPublish: boolean
  mergeJsonArrays: string[]
}[]

export type ResolverFunc = (
  params: any,
  sanitizedComponent: any
) => Promise<string>
