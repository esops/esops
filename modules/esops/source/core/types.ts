/**
 * ## Primitives
 */

export type Path = string
export type URL = Path
export type TemplateProps = {}
export type CWD = Path

export type TemporaryTemplatePath = Path
export type TempGenerationPath = Path

export type NetworkUrl = URL
export type NetworkWithProps = [NetworkUrl, TemplateProps]
export type NetworkOption = NetworkUrl | NetworkWithProps
export type NetworkOptions = NetworkOption[] | NetworkUrl
export type NetworkParams = {
  cwd: CWD
  opts: NetworkOptions
}

export type LocalPath = Path
export type LocalWithProps = [LocalPath, TemplateProps]
export type LocalOption = LocalPath | LocalWithProps
export type LocalOptions = LocalOption[] | LocalPath
export type LocalOptionsWithProps = LocalWithProps[]
export type LocalParams = {
  cwd: CWD
  opts: LocalOptions
}

export type TemplatePath = LocalPath | TemporaryTemplatePath | NetworkUrl
export type PathList = TemplatePath[]
export type WithProps = LocalWithProps | NetworkWithProps
export type OptionsWithProps = WithProps[]
export type Option = Path | [Path, TemplateProps]
export type Options = Path | Option[]
export type Params = LocalParams | NetworkParams

/**
 * ## Resolve (Network + FS)
 */
export type ResolverOptions = {
  cwd: CWD
  opts: Options
}

export type Config = {
  cwd: CWD
}
export type Resolve = (
  opts: Options,
  config: Config
) => Promise<LocalOptionsWithProps>

export type convertNetworkPathsToLocalPath = (
  options: Options
) => Promise<LocalOptions>

export interface Resolver {
  default: Resolve
  // recursiveStackResolver: (stackManifest: any) => any
}

/**
 * ## Parse (JS)
 */
export type ParserOptions = {
  cwd: CWD
  opts: LocalOptionsWithProps
}

export type Parser = (options: ParserOptions) => Promise<CopyManifest>

export type ConvertPathsToCopyManifest = (paths: Path[]) => CopyManifest

// export type PatchWhitelist = [
//   ['.json.template', 'RENDER_THEN_MERGE_JSON'],
//   ['.json', 'MERGE_JSON'],
//   ['.gitignore', 'MERGE_FILE'],
//   ['.gitignore.template', 'RENDER_THEN_MERGE_FILE'],
//   ['.npmignore', 'MERGE_FILE'],
//   ['.npmignore.template', 'RENDER_THEN_MERGE_FILE'],
//   ['.md', 'RENDER_THEN_MERGE_FILE']
//   ['.template', 'RENDER_TEMPLATE'],
//   ['', 'COPY_AND_OVERRIDE']
// ]

export type Methods =
  | 'RENDER_THEN_MERGE_JSON'
  | 'MERGE_JSON'
  | 'MERGE_FILE'
  | 'RENDER_THEN_MERGE_FILE'

export type Copies = {
  outputDir: CWD
  templateDir: TemplatePath
  relativePath: Path
  method: Methods
}
export type CopyManifest = {
  options: LocalOptions
  paths: Copies[]
}

export interface Parsers {
  default: Parser
  convertPathsToCopyManifest: ConvertPathsToCopyManifest
}

/**
 * ## Generate (FS)
 */
export type Generate = (manifest: CopyManifest) => Promise<boolean>

export type GenerateTemporaryFiles = (
  copyManifest: CopyManifest
) => Promise<TempGenerationPath>

export type ForceCopyFiles = (
  TempGenerationPath,
  DestinationPath
) => Promise<boolean>

export interface Generator {
  default: Generate
  generateTmp: GenerateTemporaryFiles
  forceCopy: ForceCopyFiles
}

/**
 * ## Run (JS)
 */

export type EsopsRun = (cwd: CWD, options?: Options) => Promise<void>

export interface Run {
  default: EsopsRun
}

/**
 * ## Bin (Process)
 */
export type EsopBin = () => Promise<void>
