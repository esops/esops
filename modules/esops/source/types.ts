type Path = string

type SourcePath = Path
type TemplatePath = Path
type TempPath = Path
type ConsumerOptions = {}
type Copy = {}
type CopyManifest = Copy[]

export interface features {
  default: (
    cwd: SourcePath,
    template?: TemplatePath,
    opts?: ConsumerOptions
  ) => Promise<void>
  resolve: (
    cwd: SourcePath,
    template?: TemplatePath,
    opts?: ConsumerOptions
  ) => Promise<{src: SourcePath; template: TemplatePath; opts: ConsumerOptions}>
  fetchTemplate: (cwd: SourcePath) => Promise<SourcePath>
  // recursiveStackResolver: (stackManifest: any) => any
  parse: (paths: Path[]) => CopyManifest
  convertPathsToCopyManifest: (paths: Path[]) => CopyManifest
  // validate: (templatePath: string) => boolean,
  generate
  generateTmp: (copyManifest: CopyManifest) => Promise<TempPath>
  forceCopy: (TempPath, DestinationPath) => Promise<boolean>
}
