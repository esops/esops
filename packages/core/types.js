/*
# Requirements

*/

export type Compiler {

}

export type Publisher {

}

// These are the main actions that a user can do after selecting their preferred shipper.
export type ShipperActions {
    // SHIPPER_LEARN_OPERATING_SYSTEM: 'SHIPPER_LEARN_OPERATING_SYSTEM',
    // SHIPPER_SETUPASSIST_OPERATING_SYSTEM: 'SHIPPER_LEARN_OPERATING_SYSTEM',

    // SHIPPER_LEARN_TERMINAL: 'SHIPPER_LEARN_TERMINAL',
    // SHIPPER_SETUPASSIST_TERMINAL: 'SHIPPER_SETUPASSIST_TERMINAL',

    // SHIPPER_LEARN_GIT: 'SHIPPER_LEARN_GIT',
    // SHIPPER_SETUPASSIST_GIT: 'SHIPPER_SETUP_GIT',

    // SHIPPER_LEARN_GIT_STAGES
    // SHIPPER_SETUPASSIST_GIT_STAGES
    // SHIPPER_CREATE_GIT_STAGES
    // SHIPPER_UPDATE_GIT_STAGE:
    // SHIPPER_GIT_SYNC_STAGE: [],
    // SHIPPER_GIT_FIXMERGECONFLICT:
    // SHIPPER_GIT_RESET_STAGE:

    // SHIPPER_LEARN_TARGETS: 'SHIPPER_LEARN_TARGETS',
    // SHIPPER_SETUPASSIST_TARGETS: 'SHIPPER_SETUPASSIST_TARGETS',
    // SHIPPER_UPDATE_TARGETS: 'SHIPPER_UPDATE_TARGETS',

    // SHIPPER_LEARN_PUBLISHER: 'SHIPPER_LEARN_PUBLISHER',
    // SHIPPER_SETUPASSIST_PUBSLIHER: 'SHIPPER_SETUPASSIST_PUBLISHER',
    // SHIPPER_UPDATE_PUBLISHER: 'SHIPPER_UPDATE_PUBLISHER'

    // SHIPPER_LEARN_CI: 'SHIPPER_LEARN_CI',
    // SHIPPER_SETUPASSIST_CI: 'SHIPPER_SETUPASSIST_CI',
    // SHIPPER_UPDATE_CI: 'SHIPPER_UPDATE_CI',
    // SHIPPER_PUBLISH_CI: 'SHIPPER_PUBLISH_CI',

    // SHIPPER_LEARN_LANGUAGES: 'SHIPPER_LEARN_LANGUAGES',
    // SHIPPER_SETUPASSIST_LANGUAGES: 'SHIPPER_SETUPASSIST_LANGUAGES',
    // SHIPPER_UPDATE_LANGUAGES: 'SHIPPER_UPDATE_LANGUAGES',

    // SHIPPER_LEARN_DESIGN_PATTERNS: 'SHIPPER_LEARN_EDITOR',
    // SHIPPER_SETUPASSIST_DESIGN_PATTERNS: 'SHIPPER_LEARN_EDITOR',
    // SHIPPER_UPDATE_DESIGN_PATTERNS: 'SHIPPER_LEARN_EDITOR',

    // SHIPPER_LEARN_EDITOR: 'SHIPPER_LEARN_EDITOR',
    // SHIPPER_SETUPASSIST_EDITOR: 'SHIPPER_SETUPASSIST_EDITOR',

    // SHIPPER_LEARN_VALIDATOR: 'SHIPPER_LEARN_VALIDATOR',
    // SHIPPER_SETUPASSIST_VALIDATOR: 'SHIPPER_SETUP_VALIDATOR',
    // SHIPPER_UPDATE_VALIDATORS: 'SHIPPER_UPDATE_VALIDATORS',

    // SHIPPER_LEARN_TESTS: 'SHIPPER_LEARN_TESTS',
    // SHIPPER_SETUPASSIST_TESTS: 'SHIPPER_SETUPASSIST_TESTS',
    // SHIPPER_UPDATE_TESTS: 'SHIPPER_UPDATE_TESTS',
}

export type Shipper {
  process: 'yarn-gitflow',
  availableProcesses: [],
  updater: 'semantic-release',
  availableUpdater: [],
  validator: any,
  test: any,
  compiler: any,
  docs: any,
}

export type Communications {
  brand: string,
  goal: string,
  description: string,
}