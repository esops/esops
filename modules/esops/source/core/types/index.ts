/**
 * ## Primitives
 */
export type Path = string
export type SourcePath = Path
export type LocalTemplatePath = Path
export type TemporaryTemplatePath = Path
export type TemplatePath = LocalTemplatePath | TemporaryTemplatePath
export type TemplateUrl = string
export type TempGenerationPath = Path

export type TemplateOptions = {}
export type EsopsManifest = TemplateUrl | [TemplateUrl, TemplateOptions]

export type Drivers = {
  fs?: any
  http?: any
  logger?: any
}
export type InstallDrivers = (any) => ResolverOptions

/**
 * ## Main
 */
export type EsopsRun = (
  cwd: SourcePath,
  template?: TemplateUrl,
  options?: TemplateOptions
) => Promise<void>

/**
 * ## Resolve
 */
export type ResolverOptions = {
  cwd: SourcePath
  template?: TemplateUrl
  options?: TemplateOptions
  drivers: Drivers
}
export type Resolve = (options: ResolverOptions) => Promise<ParserOpts>

export type FindEsopsManifest = (cwd: SourcePath) => Promise<EsopsManifest>

export type FetchTemplate = (cwd: TemplateUrl) => Promise<TemporaryTemplatePath>

/**
 * ## Parse
 */
export type ParserOpts = {
  source: SourcePath
  template: TemplatePath
  options: TemplateOptions
  drivers: Drivers
}
export type Parser = (options: ParserOpts) => Promise<GeneratorOpts>

export type ConvertPathsToCopyManifest = (paths: Path[]) => CopyManifest

export type PatchWhitelist = [
  ['.json.template', 'RENDER_THEN_MERGE_JSON'],
  ['.json', 'MERGE_JSON'],
  ['.gitignore', 'MERGE_FILE'],
  ['.gitignore.template', 'RENDER_THEN_MERGE_FILE'],
  ['.npmignore', 'MERGE_FILE'],
  ['.npmignore.template', 'RENDER_THEN_MERGE_FILE'],
  ['.template', 'RENDER_TEMPLATE'],
  ['', 'COPY_AND_OVERRIDE']
]

export type Methods =
  | 'RENDER_THEN_MERGE_JSON'
  | 'MERGE_JSON'
  | 'MERGE_FILE'
  | 'RENDER_THEN_MERGE_FILE'

export type Copies = {
  outputDir: SourcePath
  templateDir: TemplatePath
  relativePath: Path
  method: Methods
}
export type CopyManifest = {
  options: TemplateOptions
  paths: Copies[]
}

/**
 * ## Generate
 */
export type GeneratorOpts = {
  drivers: Drivers
  manifest: CopyManifest
}
export type Generate = (options: CopyManifest) => Promise<boolean>

export type GenerateTemporaryFiles = (
  copyManifest: CopyManifest
) => Promise<TempGenerationPath>

export type ForceCopyFiles = (
  TempGenerationPath,
  DestinationPath
) => Promise<boolean>

/**
 * ## Exports
 */

export interface Resolver {
  default: Resolve
  findEsopsManifest: FindEsopsManifest
  fetchTemplate: FetchTemplate
  // recursiveStackResolver: (stackManifest: any) => any
}
export interface Parsers {
  default: Parser
  convertPathsToCopyManifest: ConvertPathsToCopyManifest
}

export interface Generator {
  default: Generate
  generateTmp: GenerateTemporaryFiles
  forceCopy: ForceCopyFiles
}

export interface Run {
  default: EsopsRun
}
