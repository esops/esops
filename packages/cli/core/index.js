import conf from 'conf';

const connected = (callback) => {
  callback(features);
};

export default () => {
  const features = {
    storage
  }

  return (callback) => {
    callback(features)
  }
}