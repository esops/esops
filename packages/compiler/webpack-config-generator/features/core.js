/* eslint-disable import/prefer-default-export */

import webpack from 'webpack';
import CleanWebpackPlugin from 'clean-webpack-plugin';

export const paths = ({ buildPath, buildDir, entryPath, libraryName }) => ({
  entry: {
    main: ['babel-polyfill', entryPath].map(require.resolve)
  },

  output: {
    filename: `${libraryName}.[name].[chunkhash].js`,
    library: [libraryName],
    libraryTarget: 'umd',
    path: buildPath,
  },

  devServer: {
    contentBase: buildDir,
  },

  // Paths that can resolve in the require/import statements
  resolve: {
    extensions: ['.js', '.json', '.jsx', '.ts', '.tsx'],
  },

  // Run the webpack that is within this packages context.
  context: __dirname,
});

export const plugins = ({ buildPath }) => ([

  new CleanWebpackPlugin([buildPath], { allowExternal: true }),

  new webpack.optimize.CommonsChunkPlugin({
    name: 'vendor',
    minChunks: Infinity,
  }),

]);
