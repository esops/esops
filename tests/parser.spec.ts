import {
  parsedToGeneratorManifest,
  parseWorkingDirectory,
  resolve
} from '../parser/parse'
import {MOCK_STACKS, MOCK_TEMPLATES} from '../core/examples'
import {ParsedStack, LocalWithProps} from '../core/types'
import * as path from 'path'

import {withSnapshots} from './test-utils/withSnapshots'
import {scrubAbsolutePathFromString, prettyJSON} from './test-utils/fs-utils'

const describe = withSnapshots(__filename)

describe('parseWorkingDirectory()', async assert => {
  const stack: LocalWithProps = [
    path.join(MOCK_STACKS['extendable-file'], 'infrastructure-local'),
    {}
  ]
  const cwd = stack[0]
  const actual = await parseWorkingDirectory(stack)
  const snap = scrubAbsolutePathFromString(prettyJSON(actual), cwd)
  assert({
    given: 'single FS path',
    should: 'return fs path as LocalStackWithProps',
    snap
  })
})

describe('parsedToGeneratorManifest()', async assert => {
  const expectedRelativePaths = [
    '.vscode/settings.json',
    'src/stores/stores-architecture.md',
    'tsconfig.json'
  ]

  const config = {cwd: MOCK_TEMPLATES.basic}
  const resolved = await resolve('./', config)
  const parsed = parsedToGeneratorManifest(resolved, config)

  assert({
    given: path.basename(config.cwd),
    should: `have expected length of ${expectedRelativePaths.length}`,
    expected: expectedRelativePaths.length,
    actual: parsed.length
  })

  parsed.forEach(manifest => {
    assert({
      given: `${manifest.relativePath} in  ${path.basename(config.cwd)}`,
      should: 'exist',
      expected: true,
      actual: expectedRelativePaths.indexOf(manifest.relativePath) > -1
    })

    assert({
      given: `${manifest.relativePath}`,
      should: 'have cwd',
      expected: MOCK_TEMPLATES.basic,
      actual: manifest.cwd
    })
  })
})

describe('resolver()', async assert => {
  const config = {
    cwd: MOCK_STACKS['basic-ignore-files']
  }

  assert({
    given: 'single FS path',
    should: 'return fs path as LocalStackWithProps',
    expected: [[config.cwd + '/', {}]],
    actual: await resolve('./', config)
  })

  assert({
    given: 'single FS path as array',
    should: 'return fs path as LocalStackWithProps',
    expected: [[config.cwd + '/', {}]],
    actual: await resolve(['./'], config)
  })

  assert({
    given: 'single FS path with props',
    should: 'return fs path as LocalStackWithProps',
    expected: [[config.cwd + '/', {}]],
    actual: await resolve([['./', {}]], config)
  })
})
