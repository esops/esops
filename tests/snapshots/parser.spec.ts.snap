module.exports[`Describe parseWorkingDirectory() | Given single FS path: should return fs path as LocalStackWithProps`] = `

{
  "directory": "/current/working/directory",
  "cwd": "/current/working/directory",
  "toggles": {},
  "files": [
    "/current/working/directory/esops.json",
    "/current/working/directory/package.json"
  ],
  "opts": [
    "../infrastructure-external"
  ],
  "stack": [
    "../infrastructure-external"
  ],
  "destination": "/current/working/directory",
  "compose": [
    "../infrastructure-external"
  ]
}`

